File: Samples/DeviceEnumerationAndPairing/js/js/scenario4_snapshot.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var DevEnum = Windows.Devices.Enumeration;
    var resultCollection = new WinJS.Binding.List([]);
    var resultsListView;

    var page = WinJS.UI.Pages.define("/html/scenario4_snapshot.html", {
        ready: function (element, options) {

            // Setup beforeNavigate event
            WinJS.Navigation.addEventListener("beforenavigate", onLeavingPage);

            // Hook up button event handlers
            document.getElementById("findButton").addEventListener("click", findButtonClick, false);

            // Hook up result list data binding
            resultsListView = element.querySelector("#resultsListView").winControl;
            resultsListView.itemDataSource = resultCollection.dataSource;

            // Manually bind selector options
            DisplayHelpers.findAllAsyncSelectors.forEach(function each(item) {
                var option = document.createElement("option");
                option.textContent = item.displayName;
                selectorComboBox.appendChild(option);
            });

            // Process any data bindings
            WinJS.UI.processAll();
        }
    });

    function onLeavingPage() {
        WinJS.Navigation.removeEventListener("beforenavigate", onLeavingPage);
        resultCollection.splice(0, resultCollection.length);
    }

    function findButtonClick() {
        var deviceInfoCollection = null;

        findButton.disabled = true;
        resultCollection.splice(0, resultCollection.length);

        var selectedItem = DisplayHelpers.findAllAsyncSelectors.getAt(selectorComboBox.selectedIndex);

        WinJS.log && WinJS.log("FindAllAsync operation started...", "sample", "status");

        if (null != selectedItem.deviceClassSelector) {
            // If the a pre-canned device class selector was chosen, call the DeviceClass overload
            DevEnum.DeviceInformation.findAllAsync(selectedItem.deviceClassSelector).done(findAllComplete, findAllError);
        }
        else {
            // Use AQS string selector from dynamic call to a device api's GetDeviceSelector call
            DevEnum.DeviceInformation.findAllAsync(
                selectedItem.selector,
                null // don't request additional properties for this sample
                ).done(findAllComplete, findAllError);
        }

        findButton.disabled = false;
    }

    function findAllComplete(deviceInfoCollection) {

        WinJS.log && WinJS.log("FindAllAsync operation completed. " + deviceInfoCollection.length + " devices found.", "sample", "status");

        for (var i = 0; i < deviceInfoCollection.length; i++) {
            var deviceInfo = deviceInfoCollection.getAt(i);

            // For simplicity, just creating a new "display object" on the fly since databinding directly with deviceInfo from
            // the callback doesn't work. 
            resultCollection.push(new DisplayHelpers.DeviceInformationDisplay(deviceInfo));
        }
    }

    function findAllError(e) {
        WinJS.log && WinJS.log("FindAllAsync operation failed. error = " + e.message, "sample", "error");
    }

})();

File: Samples/DeviceEnumerationAndPairing/cppwinrt/Scenario4_Snapshot.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario4_Snapshot.h"
#include "SampleConfiguration.h"
#include "DeviceHelpers.h"
#include "Scenario4_Snapshot.g.cpp"

using namespace winrt;
using namespace winrt::Windows::Devices::Enumeration;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario4_Snapshot::Scenario4_Snapshot()
    {
        InitializeComponent();
    }

    void Scenario4_Snapshot::OnNavigatedTo(NavigationEventArgs const&)
    {
        resultsListView().ItemsSource(resultCollection);

        selectorComboBox().ItemsSource(DeviceSelectorChoices::FindAllAsyncSelectors());
        selectorComboBox().SelectedIndex(0);
    }

    fire_and_forget Scenario4_Snapshot::FindButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        findButton().IsEnabled(false);
        resultCollection.Clear();

        // First get the device selector chosen by the UI.
        IInspectable item = selectorComboBox().SelectedItem();
        DeviceSelectorInfo* deviceSelectorInfo = FromInspectable<DeviceSelectorInfo>(item);

        rootPage.NotifyUser(L"FindAllAsync operation started...", NotifyType::StatusMessage);

        DeviceInformationCollection deviceInfoCollection{ nullptr };

        if (deviceSelectorInfo->Selector().empty())
        {
            // If a pre-canned device class selector was chosen, call the DeviceClass overload
            deviceInfoCollection = co_await DeviceInformation::FindAllAsync(deviceSelectorInfo->DeviceClassSelector());
        }
        else if (deviceSelectorInfo->Kind() == DeviceInformationKind::Unknown)
        {
            // Use AQS string selector from dynamic call to a device api's GetDeviceSelector call
            // Kind will be determined by the selector
            deviceInfoCollection = co_await DeviceInformation::FindAllAsync(
                deviceSelectorInfo->Selector(),
                nullptr // don't request additional properties for this sample
            );
        }
        else
        {
            // Kind is specified in the selector info
            deviceInfoCollection = co_await DeviceInformation::FindAllAsync(
                deviceSelectorInfo->Selector(),
                nullptr, // don't request additional properties for this sample
                deviceSelectorInfo->Kind());
        }

        rootPage.NotifyUser(
            L"FindAllAsync operation completed. " + to_hstring(deviceInfoCollection.Size()) + L" devices found.",
            NotifyType::StatusMessage);

        for (DeviceInformation const& deviceInfo : deviceInfoCollection)
        {
            resultCollection.Append(make<DeviceInformationDisplay>(deviceInfo));
        }

        findButton().IsEnabled(true);
    }
}
